### PR TITLE
76 test shap

### PR DESCRIPTION
**Pull request by geek-yang**
Add three notebooks for studying the SHAP explainer.

- DeepLiftSHAP on binary MNIST dataset
- DeepLiftSHAP on geometric shapes dataset
- GradientSHAP on binary MNIST dataset

DeepLift and Integrated Gradients (IG) approaches are also tested in these notebooks, since SHAP algorithms are based on these methods.

A few typos are corrected in `MNIST_exploration.ipynb `, `generate_model_all.ipynb`, `generate_model.ipynb`, which are mostly about path.

### CODE REVIEW

File: example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb [hunk 1]
```diff
@@ -0,0 +1,491 @@
+{
```
Comment (loostrum) [inline example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb:1]: <p>There is no dependency on tensorflow, so that could be removed from the Library list</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dianna-ai/dianna-exploration/pull/105/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='0'/>

Comment (loostrum) [inline example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #10.    #import torch.onnx
</pre><p>remove commented import?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dianna-ai/dianna-exploration/pull/105/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='1'/>

Comment (loostrum) [inline example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #6.    output_path = '/mnt/d/NLeSC/DIANNA/codebase/dianna/example_data/xai_method_study'
</pre><p>The output path is not used in this notebook I think?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dianna-ai/dianna-exploration/pull/105/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='3'/>

Comment (loostrum) [inline example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb:1]: <p>This refers to MNIST instead of geometric shapes</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dianna-ai/dianna-exploration/pull/105/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='7'/>

Comment (loostrum) [inline example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb:1]: <p>I'm not sure the URL to ONNX is relevant here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dianna-ai/dianna-exploration/pull/105/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='9'/>

Comment (loostrum) [inline example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb:1]: <p>Perhaps remove commented code in this cell (and the next one)?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dianna-ai/dianna-exploration/pull/105/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='12'/>

Comment (loostrum) [inline example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb:1]: <p>Would there be a way to show the original images in grayscale here? Not important, but I think it looks a bit nicer. Perhaps it can't be done easily, the cmap option only seems to change the attribution plot, not the original.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dianna-ai/dianna-exploration/pull/105/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='14'/>

File: example_data/xai_method_study/SHAP/DeepLiftSHAP_mnist_binary.ipynb [hunk 2]
```diff
@@ -0,0 +1,498 @@
+{
```
Comment (loostrum) [inline example_data/xai_method_study/SHAP/DeepLiftSHAP_mnist_binary.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #10.    #import torch.onnx
</pre><p>Remove commented import?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dianna-ai/dianna-exploration/pull/105/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='1'/>

Comment (loostrum) [inline example_data/xai_method_study/SHAP/DeepLiftSHAP_mnist_binary.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #6.    output_path = '/mnt/d/NLeSC/DIANNA/codebase/dianna/example_data/xai_method_study'
</pre><p>Output path is not used in this notebook I think.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dianna-ai/dianna-exploration/pull/105/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='3'/>

Comment (loostrum) [inline example_data/xai_method_study/SHAP/DeepLiftSHAP_mnist_binary.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #6.    # display attraibution map
</pre><p>Remove commented code in this cell and the next?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dianna-ai/dianna-exploration/pull/105/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='13'/>

Comment (loostrum) [inline example_data/xai_method_study/SHAP/DeepLiftSHAP_mnist_binary.ipynb:1]: <p>I like these plots! Really interesting to see the 0 and 1 in the attribution plots. As for the geometric shapes notebook: it would be cool if the original images in this plot could be grayscale.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dianna-ai/dianna-exploration/pull/105/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='15'/>

File: example_data/xai_method_study/SHAP/GradientSHAP_mnist_binary.ipynb [hunk 3]
```diff
@@ -0,0 +1,429 @@
+{
```
Comment (loostrum) [inline example_data/xai_method_study/SHAP/GradientSHAP_mnist_binary.ipynb:1]: <p>No dependency on tensorflow so it could be removed from the Library list</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dianna-ai/dianna-exploration/pull/105/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='0'/>

Comment (loostrum) [inline example_data/xai_method_study/SHAP/GradientSHAP_mnist_binary.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #10.    import torch.onnx
</pre><p>This line is commented in the other notebooks and I think it's not used here either, so it could be removed</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dianna-ai/dianna-exploration/pull/105/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='1'/>

Comment (loostrum) [inline example_data/xai_method_study/SHAP/GradientSHAP_mnist_binary.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #6.    output_path = '/mnt/d/NLeSC/DIANNA/codebase/dianna/example_data/xai_method_study'
</pre><p>Output path is not used?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dianna-ai/dianna-exploration/pull/105/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='3'/>

Comment (loostrum) [inline example_data/xai_method_study/SHAP/GradientSHAP_mnist_binary.ipynb:1]: <p>Again, grayscale colourmap for the original images would be nice, but may not be doable easily.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/dianna-ai/dianna-exploration/pull/105/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='RESOLVED' data-cellIndex='16'/>

**Review by loostrum — APPROVED** [review]
Very nice! I've added some minor comments, but overall it looks good to me. The 3 notebooks are similar enough so I reviewed all of then, but there are some duplicate comments. I started with the geometric shapes one, so that's perhaps the easiest place to start. 

One overall comment: with the `exist_ok=True` flag in `os.makedirs`, you don't need to check if the directory exists beforehand anymore. This applies to several notebooks, but it doesn't matter too much either of course. The code works fine!

File: example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb (hunk 1 above)
Comment (geek-yang) [reply example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb:1]: <p>Corrected!</p>

Comment (geek-yang) [reply example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb:1]: <p>Removed!</p>

Comment (geek-yang) [reply example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb:1]: <p>Removed.</p>

Comment (geek-yang) [reply example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb:1]: <p>Sorry this notebook was created based on the one with binary MNIST (guess you notice it already ^^). I forgot to change those titles. Now it is fixed.</p>

Comment (geek-yang) [reply example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb:1]: <p>I tested some ONNX models with captum. Forgot to remove this line. It is deleted now.</p>

Comment (geek-yang) [reply example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb:1]: <p>Done!</p>

Comment (geek-yang) [reply example_data/xai_method_study/SHAP/DeepLiftSHAP_geometric_shapes.ipynb:1]: <p>This is a built-in function with captum. It is not that easy to change. I suggest we stick to this one, although I do agree that in grayscale it will be nice.</p>

File: example_data/xai_method_study/SHAP/DeepLiftSHAP_mnist_binary.ipynb (hunk 2 above)
Comment (geek-yang) [reply example_data/xai_method_study/SHAP/DeepLiftSHAP_mnist_binary.ipynb:1]: <p>Corrected!</p>

Comment (geek-yang) [reply example_data/xai_method_study/SHAP/DeepLiftSHAP_mnist_binary.ipynb:1]: <p>Fixed.</p>

Comment (geek-yang) [reply example_data/xai_method_study/SHAP/DeepLiftSHAP_mnist_binary.ipynb:1]: <p>Done.</p>

Comment (geek-yang) [reply example_data/xai_method_study/SHAP/DeepLiftSHAP_mnist_binary.ipynb:1]: <p>Thanks for your comment Leon! Well it will be nice indeed we can plot them in grayscale. But since we simply make sure of the built-in plot function with captum (it can easily show the overlay in a nice way, that's why we use this function).  It does not support display in grayscale. I suggest not be bothered with it and just keep it for now.</p>

File: example_data/xai_method_study/SHAP/GradientSHAP_mnist_binary.ipynb (hunk 3 above)
Comment (geek-yang) [reply example_data/xai_method_study/SHAP/GradientSHAP_mnist_binary.ipynb:1]: <p>Fixed.</p>

Comment (geek-yang) [reply example_data/xai_method_study/SHAP/GradientSHAP_mnist_binary.ipynb:1]: <p>Fixed.</p>

Comment (geek-yang) [reply example_data/xai_method_study/SHAP/GradientSHAP_mnist_binary.ipynb:1]: <p>Still...Let's skip this one to make life easier. But thanks a lot for your comment ^^!</p>

**Comment by geek-yang** [comment]
Thanks Leon, for the review and all the comments! I have fixed all the issues and now it can be merged!
